### ISSUE TITLE
trg_mask generate problem

### ISSUE THREAD

**Issue by tongye98** [joeynmt/joeynmt#208]
the `trg_mask` is generated in the Batch class:
`self.trg_mask: Tensor = (self.trg != pad_index).unsqueeze(1)`
[https://github.com/joeynmt/joeynmt/blob/0f57e93604b58cefadf48583f47a322819c59850/joeynmt/batch.py#L70](url)

```
            # trg_input is used for teacher forcing, last one is cut off
            self.trg_input: Tensor = trg[:, :-1]  # shape (batch_size, seq_length)
            self.trg_length: Tensor = trg_length - 1
            # trg is used for loss computation, shifted by one since BOS
            self.trg: Tensor = trg[:, 1:]  # shape (batch_size, seq_length)
            # we exclude the padded areas (and blank areas) from the loss computation
            self.trg_mask: Tensor = (self.trg != pad_index).unsqueeze(1)
```

I am confused about why use `self.trg` (used as the target ground truth in my understanding)  not the `self.trg_input` to generate the `trg_mask` ?

**Comment by may-** [comment]
Hi @tongye98 ,

Thank you for your question!
TLDR; because `trg_mask` is used for loss computation.

---

The model predicts auto-regressively, which means the model takes previous tokens as input.
Let's say the model processes the sequence `<s> hallo welt </s>` in src side (german) and `<s> hello world </s>` in trg side (english).

`self.trg_input` holds the sequences with BOS `<s>` prepended, i.e. `<s> hello world`.

In the first timestep, the model takes the src sequence`<s> hallo welt </s>` and trg start token `<s>` as input, and tries to predict next trg token `hello`. Assume that the model wrongly predicted the first token, i.e. `hi` instead of `hello`. Still, we feed the ground truth prefix to model In the second step, `<s> hallo welt </s>` and `<s> hello` so that the model can predict the second token `world` accurately. That's why we need here one token ahead (and don't need `</s>` because we don't have the next token which the model should predict after `</s>`) for teacher forcing, and `self.trg_input` serves for this purpose. (BTW, in inference time, the model doesn't have this guidance. So, the model will predict the second token based on `<s> hallo welt </s>` and `<s> hi`, can suffer from its own mistakes.)

`self.trg` holds the sequence with EOS `</s>` appended, i.e. `hello world </s>`.

We assumed that the model wrongly predicted `hi` in the first step. We further assume that the model predicted the second token `world` correctly, but missed again the third token, i.e. `!` instead of `</s>`. Then we compute the loss between the ground truth `hello world </s>` and the prediction `hi world !`, and ignore anything after the third token in the loss computation. `self.trg_mask` serves for this masking purpose, therefore, we create `self.trg_mask` from `self.trg`, not `self.trg_input`.

 I hope it helps to understand the concept. I'm happy to receive further questions :)

**Comment by tongye98** [comment]
Hi, @may- 

Thank you for your detailed answer !
Yes, i understand and agree with what you said above. 

---

But i still have a puzzle:  `trg_mask` is also used in `trg_trg_attention` in the decoder. 
` h1, _ = self.trg_trg_att(x, x, x, mask=trg_mask)`
https://github.com/joeynmt/joeynmt/blob/0f57e93604b58cefadf48583f47a322819c59850/joeynmt/transformer_layers.py#L361

So, in the training time, if the target side sentence  is `<s> hello world </s> <pad> <pad> `  for example ( `<pad>` here for batch) , `trg_input` and `trg`  are: 
```
trg_input = <s>   hello    world    </s>    <pad> 
    trg   = hello  world   </s>     <pad>   <pad>
```
---

If the `trg_mask` is generated by `trg`,  the `trg_mask` will be  `True True True False False`.
After the `subsequent_mask`, the `final trg_mask` is 

```
True   False   False   False  False
True   True    False   False  False
True   True    True    False  False
True   True    True    False  False
True   True    True    False  False
```

If the `trg_mask` is generated by `trg_input`,  the `trg_mask` will be `True True True True False`.
After the `subsequent_mask`, the `final trg_mask` is 

```
True   False   False   False  False
True   True    False   False  False
True   True    True    False  False
True   True    True    *True*   False
True   True    True    True  False
```
Let's look the two `trg_mask` differences.  For the latter, in the `trg_trg_attention`, the token` </s>` will also attention to the token `</s>` itself, (\*True* position),  wouldn't that make more sense ?



**Addtional:**

> `trg_mask`  is used for loss computation.

But in the source code,  the input of the `loss function` is only the `logits` and `kwargs['trg']` (and `pad_index` and `smoothing` in the `__init__` of `XentLoss`), not include the `trg_mask`. This is also a little confusing to me
https://github.com/joeynmt/joeynmt/blob/0f57e93604b58cefadf48583f47a322819c59850/joeynmt/loss.py#L76

**Comment by may-** [comment]
Hi @tongye98 ,

Ah, my explanation was somewhat misleading, sorry.
First of all, in the actual loss function implemented in pytorch, they decide the masking position based on the index, not masking matrix. We specify the pad index to ignore them here:

https://github.com/joeynmt/joeynmt/blob/0f57e93604b58cefadf48583f47a322819c59850/joeynmt/loss.py#L23

And you are right, the actual usage of `trg_mask` in the code is to mask out the position not to attend. (I just meant that the overall purpose why we use `trg_mask` in the self-attention is because we want to ignore them in loss calculation.)

The important point is that the self-attention mechanism never attends to BOS,  so the masking pattern after subsequent mask below

```
True   False   False   False  False
True   True    False   False  False
True   True    True    False  False
True   True    True    False  False
True   True    True    False  False
```

actually implies:

```
hello   False   False   False  False
hello   world   False   False  False
hello   world   </s>    False  False
hello   world   </s>    False  False
hello   world   </s>    False  False
```

in our example. You know, self-attention mechanism attends to the token positions to be predicted. The start position token is always given, it's not the job of the model to predict EOS...
I recommend the article [illustrated transformer](https://jalammar.github.io/illustrated-transformer/) for further details of self-attention.

This issue https://github.com/joeynmt/joeynmt/issues/151 may help, too.

If you mean, attending to future tokens makes more sense, then yeah, there are several research in that direction., i.e. 
[Attending to Future Tokens for Bidirectional Sequence Generation](https://aclanthology.org/D19-1001/)

**Comment by may-** [comment]
Please feel free to reopen the issue, if you still have questions.
